x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java

```diff
@@ -0,0 +1,117 @@
+                    if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.EXISTING_STORE) {
+                        // we don't care what the allocation ID is since we know that these shards cannot really be stale, so we can
+                        // safely ignore the allocation ID with a forced-stale allocation
+                        iterator.updateUnassigned(shardRouting.unassignedInfo(),
+                            RecoverySource.ExistingStoreRecoverySource.FORCE_STALE_PRIMARY_INSTANCE, allocation.changes());
+                    }
+                    iterator.initialize(allocateUnassignedDecision.getTargetNode().getId(), null, 0L, allocation.changes());
+                } else {
+                    iterator.removeAndIgnore(allocateUnassignedDecision.getAllocationStatus(), allocation.changes());
+                }
+            }
+        }
+    }
+
+    private static AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation, ShardRouting shardRouting) {
+        if (isResponsibleFor(allocation, shardRouting)) {
+            Decision.Type bestDecision = Decision.Type.NO;
+            RoutingNode bestNode = null;
+            final List<NodeAllocationResult> nodeAllocationResults
+                = allocation.debugDecision() ? new ArrayList<>(allocation.routingNodes().size()) : null;
+
+            for (final RoutingNode routingNode : allocation.routingNodes()) {
+                final Decision decision = allocation.deciders().canAllocate(shardRouting, routingNode, allocation);
+                if (decision.type() == Decision.Type.YES
+                    || (decision.type() == Decision.Type.THROTTLE && bestDecision != Decision.Type.YES)) {
+                    bestDecision = decision.type();
+                    bestNode = routingNode;
+                }
+                if (nodeAllocationResults != null) {
+                    nodeAllocationResults.add(new NodeAllocationResult(routingNode.node(), null, decision));
+                }
+            }
+
+            switch (bestDecision) {
+                case YES:
+                    return AllocateUnassignedDecision.yes(bestNode.node(), null, nodeAllocationResults, false);
+                case THROTTLE:
+                    return AllocateUnassignedDecision.throttle(nodeAllocationResults);
+                case NO:
+                    return AllocateUnassignedDecision.no(UnassignedInfo.AllocationStatus.DECIDERS_NO, nodeAllocationResults);
+            }
+        }
+
+        return AllocateUnassignedDecision.NOT_TAKEN;
+    }
+
+    private static boolean isResponsibleFor(RoutingAllocation allocation, ShardRouting shardRouting) {
+        final Settings settings = allocation.metaData().getIndexSafe(shardRouting.index()).getSettings();
+        return shardRouting.unassigned() && ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(settings).equals(ALLOCATOR_NAME);
```

Comment:
If we want to name the allocator in settings, I think we should just pick the right allocator per shard from the outside?

Reply:
Yes, this needs the loop to be brought up to the top level, which I now see that we can do safely. I'll try that.